### PR TITLE
perf(behavior_velocity_planner): use faster PolygonIterator

### PR DESCRIPTION
**Pull request by maxime-clem**
Signed-off-by: Maxime CLEMENT <maxime.clement@tier4.jp>

## Description

This PR adds a faster `PolygonIterator` for the `grid_map` library.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/grid_map_utils/src/polygon_iterator.cpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+#include <functional>
+#include <utility>
+
+namespace grid_map_utils
+{
+
+std::vector<Edge> PolygonIterator::calculateSortedEdges(const grid_map::Polygon & polygon)
+{
+  std::vector<Edge> edges;
+  edges.reserve(polygon.nVertices() / 2);
```
Comment (taikitanaka3) [inline common/grid_map_utils/src/polygon_iterator.cpp:33]: This is precise number to reserve
```suggestion
  edges.reserve(polygon.nVertices());
```

File: common/grid_map_utils/src/polygon_iterator.cpp [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  // calculate map indexes between pairs of intersections Y values on each X line
+  polygon_indexes_.reserve(
+    y_intersections_per_line.size() * static_cast<size_t>(grid_map.getSize()(1)));
+  const auto from_row = start_idx(0) + from_to_row.first;
+  for (size_t i = 0; i < y_intersections_per_line.size(); ++i) {
+    const auto & y_intersections = y_intersections_per_line[i];
+    auto row = static_cast<int>(from_row + i);
+    grid_map::wrapIndexToRange(row, map_size(0));
+    for (auto y_iter = y_intersections.cbegin(); y_iter != y_intersections.cend(); ++y_iter) {
+      const auto dist_from_origin = origin.y() - *y_iter + resolution;
```
Comment (taikitanaka3) [inline common/grid_map_utils/src/polygon_iterator.cpp:143]: This line is equal to adding 1 index to from value.
```suggestion
      const auto dist_from_origin = origin.y() - *y_iter;
      const auto from_col =
        std::clamp(static_cast<int>(1.0 + dist_from_origin / resolution), 0, map_size(1) - 1);

```

**Comment by taikitanaka3** [comment]
- please make PR to grid map core too if this is merged and test result become nicer.
- please add brief description to this package

File: common/grid_map_utils/src/polygon_iterator.cpp (hunk 2 above)
Comment (maxime-clem) [reply common/grid_map_utils/src/polygon_iterator.cpp]: Thank you. I applied the change [513b347](https://github.com/autowarefoundation/autoware.universe/pull/943/commits/513b347d009c5bb3ac815ace1f17db44cfc8a5a5)

File: common/grid_map_utils/src/polygon_iterator.cpp (hunk 1 above)
Comment (maxime-clem) [reply common/grid_map_utils/src/polygon_iterator.cpp]: Thank you I applied the change: [513b347](https://github.com/autowarefoundation/autoware.universe/pull/943/commits/513b347d009c5bb3ac815ace1f17db44cfc8a5a5)

**Comment by taikitanaka3** [comment]
@maxime-clem 
I close and reopen this pr for updating ci

**Comment by taikitanaka3** [comment]
@yukkysaito @TakaHoribe
This is just PR for adding new package.
There is a test for comparing polygon iterator of grid map package and @maxime-clem 's grid map utils.
I think merge isReady. 
image

**Review by yukkysaito — APPROVED** [review]
Awesome. LGTM

**Comment by maxime-clem** [comment]
Thank you for the review.
Friday I tested a new feature making use of this faster `PolygonIterator` and did not encounter any issue.
I will merge this.

**Pull request merged**